pkg/kvstore/store/syncstore.go

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package store
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// SyncStore abstracts the operations allowing to synchronize key/value pairs
+// into a kvstore.
+type SyncStore interface {
+	// UpsertKey upserts a key/value pair into the kvstore.
+	UpsertKey(ctx context.Context, key Key) error
+
+	// DeleteLocalKey removes a key from the kvstore.
+	DeleteKey(ctx context.Context, key NamedKey) error
+
+	// Release frees all resources owned by the store, leaving all keys in the
+	// kvstore intact.
+	Release()
+}
+
+// SyncStoreBackend represents the subset kvstore.BackendOperations leveraged
+// by SyncStore implementations.
+type SyncStoreBackend interface {
+	// Update creates or updates a key.
+	Update(ctx context.Context, key string, value []byte, lease bool) error
+	// Delete deletes a key.
+	Delete(ctx context.Context, key string) error
+}
+
+// wqSyncStore implements the SyncStore interface leveraging a workqueue to
+// coalescence update/delete requests and handle retries in case of errors.
+type wqSyncStore struct {
+	backend   SyncStoreBackend
+	prefix    string
+	withLease bool
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	workqueue workqueue.RateLimitingInterface
+	state     sync.Map /* map[string][]byte --- map[NamedKey.GetKeyName()]Key.Marhal() */
+
+	log *logrus.Entry
+	wg  sync.WaitGroup
```

Comment:
Do we need this WaitGroup? I couldn't find where `Release` is used except for tests.

Reply:
I've deleted the `Release()` method and replaced it with a blocking `Run(ctx)` as suggested below.